File: node/src/gate.ts
import { createBotGateBattery, heuristicAssessment, loadTypeSafePrimitives } from './battery.js';
import {
  createClientHash,
  createPoWChallenge,
  generateChallengeHtml,
  signBotToken,
  verifyBotToken,
  verifyPoW,
  type BotCookiePayload
} from './crypto.js';
import { normalizeRequest } from './normalizer.js';
import { evaluateDecision, resolvePolicy } from './policy.js';
import type {
  AssessmentResult,
  BotGateDecision,
  BotGateOptions,
  RequestInput,
  TrafficCategory
} from './types.js';

let sdkLoaded = false;
let TypeSafeClientClass: any = null;

async function getTypeSafeClientClass() {
  if (!sdkLoaded) {
    try {
      const sdk = await import('@typesafe-ai/sdk');
      TypeSafeClientClass = sdk.TypeSafeClient;
      await loadTypeSafePrimitives();
    } catch {
      TypeSafeClientClass = null;
    }
    sdkLoaded = true;
  }
  return TypeSafeClientClass;
}

export class BotGate {
  private client: any = null;
  private clientInitAttempted = false;
  private options: BotGateOptions;
  private battery = createBotGateBattery();

  constructor(options: BotGateOptions = {}) {
    this.options = options;
  }

  private async getClient() {
    if (!this.clientInitAttempted) {
      this.clientInitAttempted = true;
      const apiKey = this.options.apiKey || process.env.TYPESAFE_API_KEY;
      if (apiKey) {
        const ClientClass = await getTypeSafeClientClass();
        if (ClientClass) {
          this.client = new ClientClass({
            apiKey,
            baseURL: this.options.endpoint || process.env.TYPESAFE_ENDPOINT,
            timeout: this.options.timeout || 15000
          });
          this.battery = createBotGateBattery();
        }
      }
    }
    return this.client;
  }

  async inspect(input: RequestInput, runtimeOptions: BotGateOptions = {}): Promise<BotGateDecision> {
    const startTime = performance.now();
    const opts = { ...this.options, ...runtimeOptions };
    const policy = resolvePolicy(opts.policy);
    const { request, context } = normalizeRequest(input);

    // Fast-path 1: Path whitelist
    if (opts.whitelistedPaths) {
      for (const pattern of opts.whitelistedPaths) {
        if (typeof pattern === 'string' && request.path === pattern) {
          return this.createAllowedDecision('Path whitelisted', startTime);
        }
        if (pattern instanceof RegExp && pattern.test(request.path)) {
          return this.createAllowedDecision('Path matched whitelist pattern', startTime);
        }
      }
    }

    // Fast-path 2: IP whitelist
    if (opts.whitelistedIps && request.ip && opts.whitelistedIps.includes(request.ip)) {
      return this.createAllowedDecision('Client IP whitelisted', startTime);
    }

    // Fast-path 3: Canary honeypot traps (advanced defense pattern)
    if (opts.honeypotPaths && opts.honeypotPaths.includes(request.path)) {
      context.isHoneypot = true;
    }

    // Fast-path 4: Stateless HMAC Cookie & Velocity Tracking (advanced defense pattern)
    const secret = opts.secretKey || process.env.BOTGATE_SECRET || 'bot-gate-default-secret-key-32b!';
    const userAgent = request.headers['user-agent'] || '';
    const clientHash = createClientHash(request.ip, userAgent);
    const rawCookie = request.cookies?.['__botgate'];
    const now = Date.now();

    let cookiePayload: BotCookiePayload | null = null;
    if (rawCookie) {
      cookiePayload = verifyBotToken(rawCookie, secret);
    }

    // Check if client previously verified a challenge and token is valid
    if (cookiePayload && cookiePayload.h === clientHash && cookiePayload.v && cookiePayload.v > now) {
      // Verified human pass
      return this.createAllowedDecision('Client passed cryptographic verification', startTime);
    }

    // Track request velocity in 10-second sliding window
    let currentCount = 1;
    let windowStart = now;
    if (cookiePayload && cookiePayload.h === clientHash) {
      if (now - cookiePayload.ws < 10_000) {
        currentCount = cookiePayload.c + 1;
        windowStart = cookiePayload.ws;
      }
    }
    context.rateCount = currentCount;

    let assessment: AssessmentResult;
    const client = await this.getClient();

    if (client && !context.isHoneypot) {
      try {
        const state = {
          request: {
            method: request.method,
            path: request.path,
            headers: request.headers,
            query: request.query,
            body: (request.body ?? null) as any,
            ip: (request.ip ?? null) as any
          },
          context: {
            missing_browser_headers: context.missingBrowserHeaders,
            suspicious_signatures: context.suspiciousSignatures,
            claims_browser: context.claimsBrowser,
            is_known_search_bot: context.isKnownSearchBot,
            header_order_anomaly: context.headerOrderAnomaly,
            shannon_entropy: context.shannonEntropy,
            rate_count: context.rateCount
          }
        };

        const response = await client.systemOne({
          model: opts.model || 'jev-latest',
          state: state as any,
          questions: this.battery as any
        });

        const answers = response.answers as any;

        const isBotProb = answers.is_bot?.noul ?? 0.1;
        const isAttackProb = answers.is_attack?.noul ?? 0.05;
        const isSpoofedProb = answers.is_spoofed?.noul ?? 0.05;

        const trafficTypeAnswer = answers.traffic_type;
        const trafficType = (trafficTypeAnswer?.choice as TrafficCategory) || 'human';
        const trafficTypeConfidence = trafficTypeAnswer?.confidence ?? 0.8;
        const trafficTypeDist = trafficTypeAnswer?.probabilities ?? {};

        const riskAnswer = answers.risk_level;
        const riskScore = riskAnswer?.score ?? 0;
        const riskConfidence = riskAnswer?.confidence ?? 0.8;
        const riskDist = riskAnswer?.probabilities ?? {};

        assessment = {
          isBotProbability: isBotProb,
          isAttackProbability: isAttackProb,
          isSpoofedProbability: isSpoofedProb,
          trafficType,
          trafficTypeConfidence,
          trafficTypeDistribution: trafficTypeDist,
          riskScore,
          riskConfidence,
          riskLevelDistribution: riskDist
        };
      } catch (err) {
        if (opts.fallback === 'block') {
          return this.createBlockedDecision('TypeSafe API unavailable (fallback: block)', startTime);
        }
        if (opts.fallback === 'allow') {
          return this.createAllowedDecision('TypeSafe API unavailable (fallback: allow)', startTime);
        }
        assessment = heuristicAssessment(request, context);
      }
    } else {
      assessment = heuristicAssessment(request, context);
    }

    const durationMs = Math.round((performance.now() - startTime) * 100) / 100;
    const decision = evaluateDecision(assessment, policy, opts, durationMs, context);

    // Prepare updated stateless cookie
    const updatedPayload: BotCookiePayload = {
      h: clientHash,
      c: currentCount,
      ws: windowStart
    };
    const newToken = signBotToken(updatedPayload, secret);
    decision.setCookieHeader = `__botgate=${newToken}; Path=/; HttpOnly; SameSite=Lax; Max-Age=3600`;

    // If challenged, attach PoW challenge payload
    if (decision.shouldChallenge) {
      const pow = createPoWChallenge(secret, 4);
      decision.challengeHtml = generateChallengeHtml(pow.seed, pow.difficulty, request.url || '/');
    }

    return decision;
  }

  middleware(middlewareOptions: BotGateOptions = {}) {
    const opts = { ...this.options, ...middlewareOptions };
    const secret = opts.secretKey || process.env.BOTGATE_SECRET || 'bot-gate-default-secret-key-32b!';

    return async (req: any, res: any, next: (err?: any) => void) => {
      try {
        // Handle PoW Challenge verification endpoint
        if (req.method === 'POST' && (req.url === '/__botgate/verify' || req.path === '/__botgate/verify')) {
          let body = req.body;
          if (typeof body === 'string') {
            try { body = JSON.parse(body); } catch {}
          }
          const { seed, nonce, isAutomated, returnUrl } = body || {};
          const isValid = !isAutomated && verifyPoW(seed, nonce, 1, secret);

          if (isValid) {
            const userAgent = req.headers['user-agent'] || '';
            const ip = req.ip || req.socket?.remoteAddress || '';
            const clientHash = createClientHash(ip, userAgent);
            // Grant 30 minutes verification
            const token = signBotToken({ h: clientHash, c: 1, ws: Date.now(), v: Date.now() + 1800_000 }, secret);
            res.setHeader('Set-Cookie', `__botgate=${token}; Path=/; HttpOnly; SameSite=Lax; Max-Age=1800`);
            return res.status(200).json({ status: 'ok', redirect: returnUrl || '/' });
          } else {
            return res.status(403).json({ error: 'Verification failed' });
          }
        }

        const decision = await this.inspect(req, opts);
        req.botGate = decision;

        if (decision.setCookieHeader) {
          res.setHeader('Set-Cookie', decision.setCookieHeader);
        }

        if (decision.shouldTarpit) {
          if (opts.onTarpit) {
            return opts.onTarpit(decision, req, res);
          }
          const delay = opts.tarpitMs || 3000;
          await new Promise((r) => setTimeout(r, delay));
          return res.status(429).json({
            error: 'Too Many Requests',
            message: 'Velocity exceeded threshold',
            action: decision.action,
            reasons: decision.reasons
          });
        }

        if (decision.shouldBlock) {
          if (opts.onBlock) {
            return opts.onBlock(decision, req, res);
          }
          return res.status(403).json({
            error: 'Forbidden',
            message: 'Access blocked by BotGate security policy',
            action: decision.action,
            reasons: decision.reasons,
            category: decision.category,
            riskScore: decision.riskScore
          });
        }

        if (decision.shouldChallenge) {
          if (opts.onChallenge) {
            return opts.onChallenge(decision, req, res);
          }
          if (opts.enablePoWChallenge !== false && decision.challengeHtml) {
            res.status(428);
            res.setHeader('Content-Type', 'text/html; charset=utf-8');
            return res.end(decision.challengeHtml);
          }
          res.setHeader('X-BotGate-Action', 'challenge');
          res.setHeader('X-BotGate-Risk', decision.riskScore.toString());
        } else {
          res.setHeader('X-BotGate-Action', 'allow');
        }

        next();
      } catch (err) {
        next(err);
      }
    };
  }

  private createAllowedDecision(reason: string, startTime: number): BotGateDecision {
    const durationMs = Math.round((performance.now() - startTime) * 100) / 100;
    return {
      action: 'allow',
      shouldBlock: false,
      shouldChallenge: false,
      shouldTarpit: false,
      isBot: false,
      isAttack: false,
      category: 'human',
      riskScore: 0,
      confidence: 1.0,
      reasons: [reason],
      assessment: {
        isBotProbability: 0,
        isAttackProbability: 0,
        isSpoofedProbability: 0,
        trafficType: 'human',
        trafficTypeConfidence: 1.0,
        trafficTypeDistribution: { human: 1.0 },
        riskScore: 0,
        riskConfidence: 1.0,
        riskLevelDistribution: { '0': 1.0 }
      },
      durationMs,
      respond: () => false
    };
  }

  private createBlockedDecision(reason: string, startTime: number): BotGateDecision {
    const durationMs = Math.round((performance.now() - startTime) * 100) / 100;
    return {
      action: 'block',
      shouldBlock: true,
      shouldChallenge: false,
      shouldTarpit: false,
      isBot: true,
      isAttack: true,
      category: 'attack',
      riskScore: 3.0,
      confidence: 1.0,
      reasons: [reason],
      assessment: {
        isBotProbability: 1.0,
        isAttackProbability: 1.0,
        isSpoofedProbability: 1.0,
        trafficType: 'attack',
        trafficTypeConfidence: 1.0,
        trafficTypeDistribution: { attack: 1.0 },
        riskScore: 3.0,
        riskConfidence: 1.0,
        riskLevelDistribution: { '3': 1.0 }
      },
      durationMs,
      respond: (res: any) => {
        if (res && typeof res.status === 'function') {
          res.status(403).json({ error: 'Forbidden', message: reason });
          return true;
        }
        return false;
      }
    };
  }
}

File: python/bot_gate/gate.py
"""Core BotGate implementation for Python."""

from __future__ import annotations

import asyncio
import os
import re
import time
from typing import Any, Callable, Pattern

from .battery import AssessmentResult, create_bot_gate_battery, heuristic_assessment
from .crypto import (
    create_client_hash,
    create_pow_challenge,
    generate_challenge_html,
    sign_bot_token,
    verify_bot_token,
)
from .normalizer import normalize_request
from .policy import BotGateDecision, GatePolicy, evaluate_decision, resolve_policy

_sdk_attempted = False
_AsyncTypeSafeClient: Any = None
_TypeSafeClient: Any = None


def _load_typesafe_sdk():
    global _sdk_attempted, _AsyncTypeSafeClient, _TypeSafeClient
    if not _sdk_attempted:
        _sdk_attempted = True
        try:
            from typesafe_sdk import AsyncTypeSafeClient, TypeSafeClient
            _AsyncTypeSafeClient = AsyncTypeSafeClient
            _TypeSafeClient = TypeSafeClient
        except ImportError:
            _AsyncTypeSafeClient = None
            _TypeSafeClient = None
    return _AsyncTypeSafeClient, _TypeSafeClient


class BotGate:
    """AI-powered bot and attack detection gate using TypeSafe System One."""

    def __init__(
        self,
        policy: str | GatePolicy | dict[str, Any] | None = None,
        api_key: str | None = None,
        endpoint: str | None = None,
        model: str = "jev-latest",
        timeout: float = 15.0,
        allow_good_bots: bool = True,
        whitelisted_paths: list[str | Pattern] | None = None,
        whitelisted_ips: list[str] | None = None,
        honeypot_paths: list[str] | None = None,
        secret_key: str | None = None,
        tarpit_ms: int = 3000,
        enable_pow_challenge: bool = True,
        fallback: str = "heuristic",  # "heuristic" | "allow" | "block"
        on_block: Callable[[BotGateDecision, Any], Any] | None = None,
        on_tarpit: Callable[[BotGateDecision, Any], Any] | None = None,
    ) -> None:
        self.policy = resolve_policy(policy)
        self.api_key = api_key or os.environ.get("TYPESAFE_API_KEY")
        self.endpoint = endpoint or os.environ.get("TYPESAFE_ENDPOINT")
        self.model = model
        self.timeout = timeout
        self.allow_good_bots = allow_good_bots
        self.whitelisted_paths = whitelisted_paths or []
        self.whitelisted_ips = whitelisted_ips or []
        self.honeypot_paths = honeypot_paths or []
        self.secret_key = secret_key or os.environ.get("BOTGATE_SECRET", "bot-gate-default-secret-key-32b!")
        self.tarpit_ms = tarpit_ms
        self.enable_pow_challenge = enable_pow_challenge
        self.fallback = fallback
        self.on_block = on_block
        self.on_tarpit = on_tarpit
        self._battery = create_bot_gate_battery()

    def _is_whitelisted(self, path: str, ip: str | None) -> tuple[bool, str]:
        for pattern in self.whitelisted_paths:
            if isinstance(pattern, str) and path == pattern:
                return True, "Path whitelisted"
            elif hasattr(pattern, "search") and pattern.search(path):
                return True, "Path matched whitelist pattern"
        if ip and self.whitelisted_ips and ip in self.whitelisted_ips:
            return True, "Client IP whitelisted"
        return False, ""

    async def inspect(
        self,
        request_input: Any,
        policy: str | GatePolicy | dict[str, Any] | None = None,
    ) -> BotGateDecision:
        """Asynchronously inspects an incoming request and returns a detection decision."""
        start = time.perf_counter()
        req, ctx = normalize_request(request_input)
        effective_policy = resolve_policy(policy) if policy else self.policy

        whitelisted, reason = self._is_whitelisted(req.path, req.ip)
        if whitelisted:
            return self._allowed_decision(reason, start)

        if self.honeypot_paths and req.path in self.honeypot_paths:
            ctx.is_honeypot = True

        # Stateless HMAC Token & Velocity Tracking (advanced defense pattern)
        user_agent = req.headers.get("user-agent", "")
        client_hash = create_client_hash(req.ip or "", user_agent)
        raw_cookie = req.cookies.get("__botgate")
        now = time.time()

        cookie_payload = verify_bot_token(raw_cookie, self.secret_key) if raw_cookie else None

        # If client previously verified a challenge and token is valid
        if cookie_payload and cookie_payload.h == client_hash and cookie_payload.v and cookie_payload.v > now:
            return self._allowed_decision("Client passed cryptographic verification", start)

        # Velocity tracking in 10-second sliding window
        current_count = 1
        window_start = now
        if cookie_payload and cookie_payload.h == client_hash:
            if now - cookie_payload.ws < 10.0:
                current_count = cookie_payload.c + 1
                window_start = cookie_payload.ws
        ctx.rate_count = current_count

        assessment: AssessmentResult
        AsyncClient, _ = _load_typesafe_sdk()

        if self.api_key and AsyncClient is not None and not ctx.is_honeypot:
            try:
                state = {
                    "request": {
                        "method": req.method,
                        "path": req.path,
                        "headers": req.headers,
                        "query": req.query,
                        "body": req.body,
                        "ip": req.ip,
                    },
                    "context": {
                        "missing_browser_headers": ctx.missing_browser_headers,
                        "suspicious_signatures": ctx.suspicious_signatures,
                        "claims_browser": ctx.claims_browser,
                        "is_known_search_bot": ctx.is_known_search_bot,
                        "header_order_anomaly": ctx.header_order_anomaly,
                        "rate_count": ctx.rate_count,
                    },
                }
                async with AsyncClient(
                    api_key=self.api_key,
                    base_url=self.endpoint,
                    timeout=self.timeout,
                ) as client:
                    resp = await client.system_one(
                        model=self.model,
                        state=state,
                        questions=self._battery,
                    )
                assessment = self._parse_response(resp)
            except Exception:
                if self.fallback == "block":
                    return self._blocked_decision("TypeSafe API unavailable (fallback: block)", start)
                elif self.fallback == "allow":
                    return self._allowed_decision("TypeSafe API unavailable (fallback: allow)", start)
                assessment = heuristic_assessment(req, ctx)
        else:
            assessment = heuristic_assessment(req, ctx)

        duration = round((time.perf_counter() - start) * 1000, 2)
        decision = evaluate_decision(
            assessment=assessment,
            policy=effective_policy,
            allow_good_bots=self.allow_good_bots,
            duration_ms=duration,
            context=ctx,
        )

        # Generate updated cookie token
        updated_token = sign_bot_token(
            {"h": client_hash, "c": current_count, "ws": window_start},
            self.secret_key,
        )
        decision.set_cookie_header = f"__botgate={updated_token}; Path=/; HttpOnly; SameSite=Lax; Max-Age=3600"

        # If challenged, attach PoW HTML challenge
        if decision.should_challenge:
            seed, diff = create_pow_challenge(self.secret_key, 4)
            decision.challenge_html = generate_challenge_html(seed, diff, req.url or "/")

        return decision

    def inspect_sync(
        self,
        request_input: Any,
        policy: str | GatePolicy | dict[str, Any] | None = None,
    ) -> BotGateDecision:
        """Synchronously inspects an incoming request and returns a detection decision."""
        start = time.perf_counter()
        req, ctx = normalize_request(request_input)
        effective_policy = resolve_policy(policy) if policy else self.policy

        whitelisted, reason = self._is_whitelisted(req.path, req.ip)
        if whitelisted:
            return self._allowed_decision(reason, start)

        if self.honeypot_paths and req.path in self.honeypot_paths:
            ctx.is_honeypot = True

        # Stateless HMAC Token & Velocity Tracking (advanced defense pattern)
        user_agent = req.headers.get("user-agent", "")
        client_hash = create_client_hash(req.ip or "", user_agent)
        raw_cookie = req.cookies.get("__botgate")
        now = time.time()

        cookie_payload = verify_bot_token(raw_cookie, self.secret_key) if raw_cookie else None

        if cookie_payload and cookie_payload.h == client_hash and cookie_payload.v and cookie_payload.v > now:
            return self._allowed_decision("Client passed cryptographic verification", start)

        current_count = 1
        window_start = now
        if cookie_payload and cookie_payload.h == client_hash:
            if now - cookie_payload.ws < 10.0:
                current_count = cookie_payload.c + 1
                window_start = cookie_payload.ws
        ctx.rate_count = current_count

        assessment: AssessmentResult
        _, SyncClient = _load_typesafe_sdk()

        if self.api_key and SyncClient is not None and not ctx.is_honeypot:
            try:
                state = {
                    "request": {
                        "method": req.method,
                        "path": req.path,
                        "headers": req.headers,
                        "query": req.query,
                        "body": req.body,
                        "ip": req.ip,
                    },
                    "context": {
                        "missing_browser_headers": ctx.missing_browser_headers,
                        "suspicious_signatures": ctx.suspicious_signatures,
                        "claims_browser": ctx.claims_browser,
                        "is_known_search_bot": ctx.is_known_search_bot,
                        "header_order_anomaly": ctx.header_order_anomaly,
                        "rate_count": ctx.rate_count,
                    },
                }
                with SyncClient(
                    api_key=self.api_key,
                    base_url=self.endpoint,
                    timeout=self.timeout,
                ) as client:
                    resp = client.system_one(
                        model=self.model,
                        state=state,
                        questions=self._battery,
                    )
                assessment = self._parse_response(resp)
            except Exception:
                if self.fallback == "block":
                    return self._blocked_decision("TypeSafe API unavailable (fallback: block)", start)
                elif self.fallback == "allow":
                    return self._allowed_decision("TypeSafe API unavailable (fallback: allow)", start)
                assessment = heuristic_assessment(req, ctx)
        else:
            assessment = heuristic_assessment(req, ctx)

        duration = round((time.perf_counter() - start) * 1000, 2)
        decision = evaluate_decision(
            assessment=assessment,
            policy=effective_policy,
            allow_good_bots=self.allow_good_bots,
            duration_ms=duration,
            context=ctx,
        )

        updated_token = sign_bot_token(
            {"h": client_hash, "c": current_count, "ws": window_start},
            self.secret_key,
        )
        decision.set_cookie_header = f"__botgate={updated_token}; Path=/; HttpOnly; SameSite=Lax; Max-Age=3600"

        if decision.should_challenge:
            seed, diff = create_pow_challenge(self.secret_key, 4)
            decision.challenge_html = generate_challenge_html(seed, diff, req.url or "/")

        return decision

    def _parse_response(self, resp: Any) -> AssessmentResult:
        answers = resp.answers

        is_bot_prob = getattr(answers.get("is_bot"), "noul", 0.1)
        is_attack_prob = getattr(answers.get("is_attack"), "noul", 0.05)
        is_spoofed_prob = getattr(answers.get("is_spoofed"), "noul", 0.05)

        traffic_answer = answers.get("traffic_type")
        traffic_type = getattr(traffic_answer, "choice", "human")
        traffic_conf = getattr(traffic_answer, "confidence", 0.8)
        traffic_dist = getattr(traffic_answer, "probabilities", {})

        risk_answer = answers.get("risk_level")
        risk_score = getattr(risk_answer, "score", 0.0)
        risk_conf = getattr(risk_answer, "confidence", 0.8)
        risk_dist = getattr(risk_answer, "probabilities", {})

        return AssessmentResult(
            is_bot_probability=is_bot_prob,
            is_attack_probability=is_attack_prob,
            is_spoofed_probability=is_spoofed_prob,
            traffic_type=traffic_type,
            traffic_type_confidence=traffic_conf,
            traffic_type_distribution=traffic_dist,
            risk_score=risk_score,
            risk_confidence=risk_conf,
            risk_level_distribution=risk_dist,
        )

    def _allowed_decision(self, reason: str, start: float) -> BotGateDecision:
        duration = round((time.perf_counter() - start) * 1000, 2)
        return BotGateDecision(
            action="allow",
            should_block=False,
            should_challenge=False,
            should_tarpit=False,
            is_bot=False,
            is_attack=False,
            category="human",
            risk_score=0.0,
            confidence=1.0,
            reasons=[reason],
            duration_ms=duration,
        )

    def _blocked_decision(self, reason: str, start: float) -> BotGateDecision:
        duration = round((time.perf_counter() - start) * 1000, 2)
        return BotGateDecision(
            action="block",
            should_block=True,
            should_challenge=False,
            should_tarpit=False,
            is_bot=True,
            is_attack=True,
            category="attack",
            risk_score=3.0,
            confidence=1.0,
            reasons=[reason],
            duration_ms=duration,
        )
